File: AircraftWar/src/main/scala/work/chiro/game/aircraft/ELiteEnemy.scala
package work.chiro.game.aircraft

import work.chiro.game.animate.{AnimateContainer, AnimateLinear, AnimateVectorType}
import work.chiro.game.application.{ImageResource, Main}
import work.chiro.game.basic.PositionType.Position
import work.chiro.game.basic.Vec2Double
import work.chiro.game.utils.{getNewFlightPosition, getTimeMills}

/**
 * 普通敌机，不可攻击
 *
 * @author chiro2001
 * @param posInit          敌机初始位置
 * @param animateContainer 动画容器
 * @param hpInit           初始血量
 */
class ELiteEnemy(posInit: Position, animateContainer: AnimateContainer[Vec2Double], hpInit: Int)
  extends AbstractAircraft(posInit, animateContainer, hpInit) {
  override def forward() = {
    super.forward()
    // 判定 y 轴向下飞行出界
    if (getLocationY >= Main.WINDOW_HEIGHT) vanish()
  }

  override def shoot() = List()

  override def getImage = ELiteEnemy.getImage

  override def create() = ELiteEnemy.create()
}

object ELiteEnemy extends ImageResource {
  override def getImageCachedPath = "images/elite.png"
  def create() = {
    val positionEnemyNew = getNewFlightPosition(ELiteEnemy.getImage.getWidth)
    new ELiteEnemy(
      positionEnemyNew, new AnimateContainer[Position](List(
        new AnimateLinear(positionEnemyNew, new Position(positionEnemyNew.getX, Main.WINDOW_HEIGHT),
          AnimateVectorType.PositionLike.id, getTimeMills, 20000)
      )), 30
    )
  }
}

File: AircraftWar/src/main/java/edu/hitsz/aircraft/ELiteEnemy.java
package edu.hitsz.aircraft;

import edu.hitsz.application.Main;
import edu.hitsz.bullet.AbstractBullet;
import edu.hitsz.bullet.EnemyBullet;

import java.util.LinkedList;

/**
 * 经营敌机
 * 不可射击
 *
 * @author hitsz
 */
public class ELiteEnemy extends AbstractAircraft {

    int shootDivider = 4;
    int shootCnt = 0;

    public ELiteEnemy(int locationX, int locationY, int speedX, int speedY, int hp) {
        super(locationX, locationY, speedX, speedY, hp);
    }

    @Override
    public void forward() {
        super.forward();
        // 判定 y 轴向下飞行出界
        if (locationY >= Main.WINDOW_HEIGHT) {
            vanish();
        }
    }

    @Override
    public LinkedList<AbstractBullet> shoot() {
        LinkedList<AbstractBullet> ret =  new LinkedList<>();
        if (shootCnt >= shootDivider) {
            ret.add(new EnemyBullet(getLocationX(), getLocationY(), 0, 10, 10));
            shootCnt = 0;
        } else {
            shootCnt += 1;
        }
        return ret;
    }

}
